pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java

```diff
@@ -209,6 +276,87 @@ public void acknowledgeMessage(List<Position> positions, AckType ackType, Map<St
+            if (this.txnID != null && this.txnID != txnID) {
+                String errorMsg = "[" + topicName + "][" + subName + "] Transaction:" + txnID.getMostSigBits() +
+                        txnID.getLeastSigBits() + " try to cumulative ack message while transaction:"
+                        + this.txnID.getMostSigBits() + this.txnID.getLeastSigBits()
+                        + " already cumulative acked messages.";
+                log.error(errorMsg);
+                throw new TransactionConflictException(errorMsg);
+            }
+
+            if (positions.size() != 1) {
+                String errorMsg = "[" + topicName + "][" + subName + "] Transaction:" + txnID.getMostSigBits() +
+                txnID.getLeastSigBits() + " invalid cumulative ack received with multiple message ids.";
+                log.error(errorMsg);
+                throw new IllegalArgumentException(errorMsg);
+            }
+
+            Position position = positions.get(0);
+
+            if (log.isDebugEnabled()) {
+                log.debug("[{}][{}] TxnID:[{}{}] Cumulative ack on {}.", topicName, subName, txnID.getMostSigBits(),
+                        txnID.getLeastSigBits(), position);
+            }
+
+             if (this.txnID == null) {
+                // Only set txnID if no transaction is doing cumulative ack.
+                TXNID_UPDATER.set(this, txnID);
+                POSITION_UPDATER.set(this, position);
+            } else if (((PositionImpl)position).compareTo((PositionImpl)this.pendingCumulativeAckMessage) > 0) {
+                // If new cumulative ack position is greater than current one, update it.
+                POSITION_UPDATER.set(this, position);
+            }
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("[{}][{}] TxnID:[{}{}] Individual acks on {}", topicName, subName, txnID.getMostSigBits(),
+                        txnID.getLeastSigBits(), positions);
+            }
+
+            ConcurrentOpenHashSet<Position> pendingAckMessage =
+                    pendingAckMessagesMap.computeIfAbsent(txnID, txn -> new ConcurrentOpenHashSet<>());
+
+            for (Position position : positions) {
+                // If try to ack message already acked by some transaction(can be itself), throw exception.
+                // Acking single message within range of cumulative ack(if exist) is considered valid operation.
+                if (this.pendingAckMessages.contains(position)) {
+                    String errorMsg = "[" + topicName + "][" + subName + "] Transaction:" + txnID.getMostSigBits() +
+                            txnID.getLeastSigBits() + " try to ack message:" + position + " already acked before.";
+                    log.error(errorMsg);
+                    throw new TransactionConflictException(errorMsg);
+                }
+
+                pendingAckMessage.add(position);
+                this.pendingAckMessages.add(position);
+            }
+        }
```

Comment:
I think we need to persistent these pending acks to cursor. so when a broker crashes, we can restore/recover those pending acks back. otherwise if a broker crashes, all the pending acks are gone.

This can probably be done in a subsequent change though. We can consider updating the description of this PR to `introduce in-memory PENDING_ACK state`.

Reply:
hmmm, you're right, I missed this part, will add in next pr.